// generic-filehandle/dist/localFile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _fs = _interopRequireDefault(require("fs"));

var _es6Promisify = require("es6-promisify");

var fsOpen = _fs.default && (0, _es6Promisify.promisify)(_fs.default.open);
var fsRead = _fs.default && (0, _es6Promisify.promisify)(_fs.default.read);
var fsFStat = _fs.default && (0, _es6Promisify.promisify)(_fs.default.fstat);
var fsReadFile = _fs.default && (0, _es6Promisify.promisify)(_fs.default.readFile);
var fsClose = _fs.default && (0, _es6Promisify.promisify)(_fs.default.close);

var LocalFile = /*#__PURE__*/function () {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  function LocalFile(source) {
    var opts = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    (0, _classCallCheck2.default)(this, LocalFile);
    (0, _defineProperty2.default)(this, "fd", void 0);
    (0, _defineProperty2.default)(this, "filename", void 0);
    this.filename = source;
  }

  (0, _createClass2.default)(LocalFile, [{
    key: "getFd",
    value: function getFd() {
      if (!this.fd) {
        this.fd = fsOpen(this.filename, 'r');
      }

      return this.fd;
    }
  }, {
    key: "read",
    value: function () {
      var _read = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(buffer) {
        var offset,
            length,
            position,
            fetchLength,
            ret,
            _args = arguments;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                offset = _args.length > 1 && _args[1] !== undefined ? _args[1] : 0;
                length = _args.length > 2 ? _args[2] : undefined;
                position = _args.length > 3 && _args[3] !== undefined ? _args[3] : 0;
                fetchLength = Math.min(buffer.length - offset, length);
                _context.t0 = fsRead;
                _context.next = 7;
                return this.getFd();

              case 7:
                _context.t1 = _context.sent;
                _context.t2 = buffer;
                _context.t3 = offset;
                _context.t4 = fetchLength;
                _context.t5 = position;
                _context.next = 14;
                return (0, _context.t0)(_context.t1, _context.t2, _context.t3, _context.t4, _context.t5);

              case 14:
                ret = _context.sent;
                return _context.abrupt("return", {
                  bytesRead: ret,
                  buffer: buffer
                });

              case 16:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function read(_x) {
        return _read.apply(this, arguments);
      }

      return read;
    }()
  }, {
    key: "readFile",
    value: function () {
      var _readFile = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee2(options) {
        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                return _context2.abrupt("return", fsReadFile(this.filename, options));

              case 1:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function readFile(_x2) {
        return _readFile.apply(this, arguments);
      }

      return readFile;
    }() // todo memoize

  }, {
    key: "stat",
    value: function () {
      var _stat = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee3() {
        return _regenerator.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.t0 = fsFStat;
                _context3.next = 3;
                return this.getFd();

              case 3:
                _context3.t1 = _context3.sent;
                return _context3.abrupt("return", (0, _context3.t0)(_context3.t1));

              case 5:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function stat() {
        return _stat.apply(this, arguments);
      }

      return stat;
    }()
  }, {
    key: "close",
    value: function () {
      var _close = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee4() {
        return _regenerator.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.t0 = fsClose;
                _context4.next = 3;
                return this.getFd();

              case 3:
                _context4.t1 = _context4.sent;
                return _context4.abrupt("return", (0, _context4.t0)(_context4.t1));

              case 5:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function close() {
        return _close.apply(this, arguments);
      }

      return close;
    }()
  }]);
  return LocalFile;
}();

exports.default = LocalFile;
